drop-transfer/src/check.rs

```diff
@@ -7,51 +7,80 @@ use tokio_util::sync::CancellationToken;
 use crate::{protocol, service::State, tasks::AliveGuard, IncomingTransfer, Transfer};
 
 pub(crate) fn spawn(
+    mut refresh_trigger: tokio::sync::watch::Receiver<()>,
     state: Arc<State>,
     xfer: Arc<IncomingTransfer>,
     logger: Logger,
     guard: AliveGuard,
     stop: CancellationToken,
-) -> tokio::task::JoinHandle<()> {
+) {
     let id = xfer.id();
-    let job = run(state, xfer, logger.clone());
 
     tokio::spawn(async move {
         let _guard = guard;
 
+        let task = async {
+            loop {
+                let cf = run(&state, xfer.clone(), &logger).await;
+
+                if cf.is_break() {
+                    break;
+                }
+
+                let _ = refresh_trigger.changed().await;
+            }
+        };
+
         tokio::select! {
             biased;
 
             _ = stop.cancelled() => {
-                debug!(logger, "Check job stop for transfer: {id}");
+                debug!(&logger, "stop checking job for: {}", id);
             },
-            _ = job => (),
+            _ = task => ()
         }
-    })
+    });
 }
 
-async fn run(state: Arc<State>, xfer: Arc<IncomingTransfer>, logger: Logger) {
-    loop {
-        tokio::time::sleep(drop_config::ALIVE_CHECK_INTERVAL).await;
+async fn run(state: &State, xfer: Arc<IncomingTransfer>, logger: &Logger) -> ControlFlow<()> {
+    let mut remove_xfer = false;
+    for delay in std::iter::once(Duration::from_secs(0)).chain(drop_config::RETRY_INTERVALS) {
+        debug!(
+            logger,
+            "Incoming transfer job started for {}, will sleep for {:?}",
+            xfer.id(),
+            delay
+        );
+        tokio::time::sleep(delay).await;
 
         if !state.transfer_manager.is_incoming_alive(xfer.id()).await {
+            remove_xfer = true;
             break;
         }
 
-        if make_request(&state, &xfer, &logger).await.is_break() {
+        if make_request(state, &xfer, logger).await.is_break() {
+            remove_xfer = true;
             break;
         }
```

Comment:
Minor: you can do something like this
```rust
        if !state.transfer_manager.is_incoming_alive(xfer.id()).await {
            return ControlFlow::Break(());
            break;
        }

        make_request(state, &xfer, logger).await.is_break()?;
```
And move the transfer removal logic to the `let task = async {...}` in the `if cf.is_break() { here }```
It would be less line of code to comprehend but you can skip this, it's OK as it is.
Note: you can use `?` on `ControlFlow` to break early with `Break`, same as for `Err()`

Reply:
I gave both comments a go and it produced a nicer code so I've commited it